File: backend/internal/models/hosted_subscription.go
package models

import (
	"time"
)

// BillingCycleType defines the allowed values for billing cycles.
type BillingCycleType string

const (
	BillingMonthly  BillingCycleType = "Monthly"
	BillingAnnually BillingCycleType = "Annually"
)

// HostedSubscription represents a subscription plan offered for sharing by a host.
// @name HostedSubscription
type HostedSubscription struct {
	ID                    uint                     `gorm:"primarykey" json:"id"`
	CreatedAt             time.Time                `json:"createdAt"`
	UpdatedAt             time.Time                `json:"updatedAt"`
	HostUserID            uint                     `gorm:"not null" json:"host_user_id"`
	User                  User                     `gorm:"foreignKey:HostUserID" json:"-"`
	SubscriptionServiceID uint                     `gorm:"not null" json:"subscription_service_id"`
	SubscriptionService   SubscriptionService      `gorm:"foreignKey:SubscriptionServiceID" json:"-"`
	SubscriptionTitle     string                   `gorm:"type:varchar(255);not null" json:"subscription_title"`
	PlanDetails           string                   `gorm:"type:text" json:"plan_details,omitempty"`
	TotalSlots            int                      `gorm:"not null" json:"total_slots"`
	CostPerCycle          float64                  `gorm:"not null" json:"cost_per_cycle"`
	BillingCycle          BillingCycleType         `gorm:"type:varchar(20);not null" json:"billing_cycle"`
	PaymentQRCodeURL      string                   `gorm:"type:text" json:"payment_qr_code_url,omitempty"`
	Description           string                   `gorm:"type:text" json:"description,omitempty"`
	Memberships           []SubscriptionMembership `gorm:"foreignKey:HostedSubscriptionID" json:"-"`
}

// CreateHostedSubscriptionRequest defines the request body for creating a new hosted subscription.
// @name CreateHostedSubscriptionRequest
type CreateHostedSubscriptionRequest struct {
	SubscriptionServiceID uint             `json:"subscription_service_id" validate:"required,gt=0"`
	SubscriptionTitle     string           `json:"subscription_title" validate:"required,min=3,max=100"`
	PlanDetails           string           `json:"plan_details,omitempty" validate:"max=255"`
	TotalSlots            int              `json:"total_slots" validate:"required,min=1,max=20"`
	CostPerCycle          float64          `json:"cost_per_cycle" validate:"required,gt=0"`
	BillingCycle          BillingCycleType `json:"billing_cycle" validate:"required,oneof=Monthly Annually"`
	PaymentQRCodeURL      string           `json:"payment_qr_code_url" validate:"omitempty,url"`
	Description           string           `json:"description,omitempty" validate:"max=1000"`
}

// HostedSubscriptionResponse is the DTO for returning hosted subscription details.
// @name HostedSubscriptionResponse
type HostedSubscriptionResponse struct {
	ID                uint             `json:"id"`
	Host              *UserResponse    `json:"host,omitempty"`
	SubscriptionTitle string           `json:"subscription_title"`
	PlanDetails       string           `json:"plan_details,omitempty"`
	TotalSlots        int              `json:"total_slots"`
	CostPerCycle      float64          `json:"cost_per_cycle"`
	BillingCycle      BillingCycleType `json:"billing_cycle"`
	PaymentQRCodeURL  string           `json:"payment_qr_code_url,omitempty"`
	Description       string           `json:"description,omitempty"`
	CreatedAt         time.Time        `json:"createdAt"`
	UpdatedAt         time.Time        `json:"updatedAt"`

	// Enriched / Calculated data
	SubscriptionServiceName string   `json:"subscription_service_name"`
	SubscriptionServiceLogo string   `json:"subscription_service_logo_url,omitempty"`
	MembersCount            int      `json:"members_count"`
	AvailableSlots          int      `json:"available_slots"`
	CostPerSlot             float64  `json:"cost_per_slot"`
	MemberAvatars           []string `json:"member_avatars,omitempty"`
}

// UserResponse is a DTO for user details included in other responses.
// @name UserResponse
type UserResponse struct {
	ID                uint    `json:"id"`
	Email             string  `json:"email"`
	FullName          string  `json:"full_name"`
	ProfilePictureURL *string `json:"profile_picture_url,omitempty"`
}

File: hubster_app/lib/models/subscriptions/hosted_subscription.dart
import 'package:equatable/equatable.dart';
import 'package:hubster_app/models/auth/user.dart';

// Enum for BillingCycleType
enum BillingCycleType {
  Monthly,
  Annually;

  static BillingCycleType fromString(String? value) {
    if (value == 'Monthly') return BillingCycleType.Monthly;
    if (value == 'Annually') return BillingCycleType.Annually;
    print(
      "Warning: Unknown BillingCycleType string '$value'. Defaulting to Monthly.",
    );
    return BillingCycleType.Monthly;
  }

  String toJsonString() {
    return name;
  }
}

class HostedSubscriptionResponse extends Equatable {
  final int id;
  final User? host;
  final String subscriptionTitle;
  final String? planDetails;
  final int totalSlots;
  final double costPerCycle;
  final BillingCycleType billingCycle;
  final String? paymentQRCodeUrl;
  final String? description;
  final DateTime createdAt;
  final DateTime updatedAt;
  final String subscriptionServiceName;
  final String? subscriptionServiceLogoUrl;
  final int membersCount;
  final int availableSlots;
  final double costPerSlot;
  final List<String>? memberAvatars;

  const HostedSubscriptionResponse({
    required this.id,
    this.host,
    required this.subscriptionTitle,
    this.planDetails,
    required this.totalSlots,
    required this.costPerCycle,
    required this.billingCycle,
    this.paymentQRCodeUrl,
    this.description,
    required this.createdAt,
    required this.updatedAt,
    required this.subscriptionServiceName,
    this.subscriptionServiceLogoUrl,
    required this.membersCount,
    required this.availableSlots,
    required this.costPerSlot,
    this.memberAvatars,
  });

  factory HostedSubscriptionResponse.fromJson(Map<String, dynamic> json) {
    print(
      "Attempting HostedSubscriptionResponse.fromJson with keys: ${json.keys.toList()}",
    );

    // Explicitly check and default non-nullable strings
    String title = "Default Title";
    if (json.containsKey('subscription_title') &&
        json['subscription_title'] is String) {
      title = json['subscription_title'];
    } else {
      print(
        "Warning: 'subscription_title' is missing or not a String. Value: ${json['subscription_title']}",
      );
    }

    String serviceName = "Default Service";
    if (json.containsKey('subscription_service_name') &&
        json['subscription_service_name'] is String) {
      serviceName = json['subscription_service_name'];
    } else {
      print(
        "Warning: 'subscription_service_name' is missing or not a String. Value: ${json['subscription_service_name']}",
      );
    }

    DateTime cAt = DateTime.fromMillisecondsSinceEpoch(0, isUtc: true);
    if (json.containsKey('createdAt') && json['createdAt'] is String) {
      try {
        cAt = DateTime.parse(json['createdAt']);
      } catch (e) {
        print("Error parsing createdAt: ${json['createdAt']}");
      }
    } else {
      print(
        "Warning: 'createdAt' is missing or not a String. Value: ${json['createdAt']}",
      );
    }

    DateTime uAt = DateTime.fromMillisecondsSinceEpoch(0, isUtc: true);
    if (json.containsKey('updatedAt') && json['updatedAt'] is String) {
      try {
        uAt = DateTime.parse(json['updatedAt']);
      } catch (e) {
        print("Error parsing updatedAt: ${json['updatedAt']}");
      }
    } else {
      print(
        "Warning: 'updatedAt' is missing or not a String. Value: ${json['updatedAt']}",
      );
    }

    BillingCycleType cycle = BillingCycleType.Monthly;
    if (json.containsKey('billing_cycle') && json['billing_cycle'] is String) {
      cycle = BillingCycleType.fromString(json['billing_cycle']);
    } else if (json['billing_cycle'] == null) {
      print("Warning: 'billing_cycle' is null. Defaulting.");
    } else {
      print(
        "Warning: 'billing_cycle' is missing or not a String. Value: ${json['billing_cycle']}",
      );
    }

    return HostedSubscriptionResponse(
      id: (json['id'] as num?)?.toInt() ?? 0,
      host:
          json.containsKey('host') &&
                  json['host'] != null &&
                  json['host'] is Map<String, dynamic>
              ? User.fromJson(json['host'] as Map<String, dynamic>)
              : null,
      subscriptionTitle: title,
      planDetails: json['plan_details'] as String?,
      totalSlots: (json['total_slots'] as num?)?.toInt() ?? 0,
      costPerCycle: (json['cost_per_cycle'] as num?)?.toDouble() ?? 0.0,
      billingCycle: cycle,
      paymentQRCodeUrl:
          json['payment_qr_code_url'] as String?, 
      description: json['description'] as String?, 
      createdAt: cAt,
      updatedAt: uAt,
      subscriptionServiceName: serviceName,
      subscriptionServiceLogoUrl:
          json['subscription_service_logo_url'] as String?, 
      membersCount: (json['members_count'] as num?)?.toInt() ?? 0,
      availableSlots: (json['available_slots'] as num?)?.toInt() ?? 0,
      costPerSlot: (json['cost_per_slot'] as num?)?.toDouble() ?? 0.0,
      memberAvatars:
          (json['member_avatars'] as List<dynamic>?)
              ?.whereType<String>()
              .toList(),
    );
  }

  @override
  List<Object?> get props => [
    id,
    host,
    subscriptionTitle,
    planDetails,
    totalSlots,
    costPerCycle,
    billingCycle,
    paymentQRCodeUrl,
    description,
    createdAt,
    updatedAt,
    subscriptionServiceName,
    subscriptionServiceLogoUrl,
    membersCount,
    availableSlots,
    costPerSlot,
    memberAvatars,
  ];
}

// DTO for creating a new hosted subscription
class CreateHostedSubscriptionRequest {
  final int subscriptionServiceId;
  final String subscriptionTitle;
  final String? planDetails;
  final int totalSlots;
  final double costPerCycle;
  final BillingCycleType billingCycle;
  final String? paymentQRCodeUrl;
  final String? description;

  CreateHostedSubscriptionRequest({
    required this.subscriptionServiceId,
    required this.subscriptionTitle,
    this.planDetails,
    required this.totalSlots,
    required this.costPerCycle,
    required this.billingCycle,
    this.paymentQRCodeUrl,
    this.description,
  });

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = {};
    data['subscription_service_id'] = subscriptionServiceId;
    data['subscription_title'] = subscriptionTitle;
    if (planDetails != null) data['plan_details'] = planDetails;
    data['total_slots'] = totalSlots;
    data['cost_per_cycle'] = costPerCycle;
    data['billing_cycle'] = billingCycle.toJsonString();
    if (paymentQRCodeUrl != null) {
      data['payment_qr_code_url'] = paymentQRCodeUrl;
    }
    if (description != null) data['description'] = description;
    return data;
  }
}
